src/compiler/checker.ts

```diff
@@ -7112,6 +7154,27 @@ namespace ts {
             }
         }
 
+        function getSubstitutionType(typeParameter: TypeParameter, substitute: Type) {
+            const result = <SubstitutionType>createType(TypeFlags.Substitution);
+            result.typeParameter = typeParameter;
+            result.substitute = substitute;
+            return result;
+        }
+
+        function getConstrainedTypeParameter(typeParameter: TypeParameter, node: Node) {

```

Comment:
getConditionallyConstrainedTypeParameter? The type parameter is not constrained in the usual sense, but substituted.

Also one line of documentation might be useful.